File: tictactoe.js
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const board = [
    [" ", " ", " "],
    [" ", " ", " "],
    [" ", " ", " "]
];

const printBoard = () => {
    console.clear();
    console.log("\n     A   B   C ");
    for (let i = 0; i < 3; i++) {
        console.log("   -------------");
        console.log(`${i + 1}  | ${board[i][0]} | ${board[i][1]} | ${board[i][2]} |`);
    }
    console.log("   -------------\n");
};

const checkWinner = () => {
    for (let i = 0; i < 3; i++) {
        if (board[i][0] === board[i][1] && board[i][1] === board[i][2] && board[i][0] !== " ") return board[i][0];
        if (board[0][i] === board[1][i] && board[1][i] === board[2][i] && board[0][i] !== " ") return board[0][i];
    }
    if (board[0][0] === board[1][1] && board[1][1] === board[2][2] && board[0][0] !== " ") return board[0][0];
    if (board[0][2] === board[1][1] && board[1][1] === board[2][0] && board[0][2] !== " ") return board[0][2];
    return board.flat().includes(" ") ? null : "Draw";
};

const playGame = (currentPlayer) => {
    printBoard();
    rl.question(`Player ${currentPlayer}, enter your move (e.g., B2) or type 'q' to quit: `, (move) => {
        if (move.toLowerCase() === "q") {
            console.log("Thanks for playing!");
            rl.close();
            return;
        }

        const col = move.charCodeAt(0) - "A".charCodeAt(0);
        const row = parseInt(move[1]) - 1;

        if (row < 0 || row > 2 || col < 0 || col > 2 || board[row][col] !== " ") {
            console.log("Invalid move, try again.");
            return playGame(currentPlayer);
        }

        board[row][col] = currentPlayer;
        const winner = checkWinner();

        if (winner) {
            printBoard();
            console.log(winner === "Draw" ? "It's a draw!" : `Player ${winner} wins!`);
            rl.close();
        } else {
            playGame(currentPlayer === "X" ? "O" : "X");
        }
    });
};

console.log("Welcome to Neo's Tic-Tac-Toe!");
playGame("X");

File: tictactoe.rs
use std::io;

fn print_board(board: &[[char; 3]; 3]) {
    println!("\n     A   B   C ");
    for (i, row) in board.iter().enumerate() {
        println!("   -------------");
        println!(" {} | {} | {} | {} |", i + 1, row[0], row[1], row[2]);
    }
    println!("   -------------\n");
}

fn check_winner(board: &[[char; 3]; 3]) -> Option<char> {
    for i in 0..3 {
        if board[i][0] == board[i][1] && board[i][1] == board[i][2] && board[i][0] != ' ' {
            return Some(board[i][0]);
        }
        if board[0][i] == board[1][i] && board[1][i] == board[2][i] && board[0][i] != ' ' {
            return Some(board[0][i]);
        }
    }
    if board[0][0] == board[1][1] && board[1][1] == board[2][2] && board[0][0] != ' ' {
        return Some(board[0][0]);
    }
    if board[0][2] == board[1][1] && board[1][1] == board[2][0] && board[0][2] != ' ' {
        return Some(board[0][2]);
    }
    if board.iter().all(|row| row.iter().all(|&cell| cell != ' ')) {
        return Some('D');
    }
    None
}

fn main() {
    let mut board = [[' '; 3]; 3];
    let mut current_player = 'X';

    println!("Welcome to Neo's Tic-Tac-Toe!");
    println!("Move using column letter and row number (e.g., B2)");
    
    loop {
        print_board(&board);
        println!("Player {}, enter your move (e.g., B2) or type 'q' to quit: ", current_player);
        
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let input = input.trim().to_uppercase();

        if input == "Q" {
            println!("Thanks for playing!");
            break;
        }

        if input.len() != 2 {
            println!("Invalid input. Use format like B2.");
            continue;
        }

        let col = match input.chars().nth(0).unwrap() {
            'A' => 0,
            'B' => 1,
            'C' => 2,
            _ => {
                println!("Invalid column. Use A, B, or C.");
                continue;
            }
        };

        let row = match input.chars().nth(1).unwrap().to_digit(10) {
            Some(n) if (1..=3).contains(&n) => (n - 1) as usize,
            _ => {
                println!("Invalid row. Use 1, 2, or 3.");
                continue;
            }
        };

        if board[row][col] != ' ' {
            println!("Cell already taken. Try again.");
            continue;
        }

        board[row][col] = current_player;

        if let Some(winner) = check_winner(&board) {
            print_board(&board);
            if winner == 'D' {
                println!("It's a draw!");
            } else {
                println!("Player {} wins!", winner);
            }
            break;
        }

        current_player = if current_player == 'X' { 'O' } else { 'X' };
    }
}
